File: Matlab_Codes/Newton_Gregory.m
%Newton Gregory mehtod 
%Written by Yusuf Hani Almoadhen 202107475

clc
clear
close all;

syms x

t = [0;1;2;3;4;5;6];
I = [176;185;194;203;212;220;229];

N = "None";
Diff1=[];
Diff2=[N];
Diff3=[N;N];
Diff4=[N;N];
Diff5=[N;N;N];
Diff6=[N;N;N];

h = t(2,1)-t(1,1);

for i=1:length(I)-1
    diff1 = I(i+1,1)-I(i,1);
    Diff1(i,1) = diff1;
end

for i=1:length(Diff1)-1
    diff2 = Diff1(i+1,1)-Diff1(i,1);
    Diff2(i+1,1) = diff2;
end

for i=2:length(Diff2)-1
    diff3 = str2num(Diff2(i+1,1))-str2num(Diff2(i,1));
    Diff3(i+1,1) = diff3;
end

for i=3:length(Diff3)-1
    diff4 = str2num(Diff3(i+1,1))-str2num(Diff3(i,1));
    Diff4(i,1) = diff4;
end

for i=3:length(Diff4)-1
    diff5 = str2num(Diff4(i+1,1))-str2num(Diff4(i,1));
    Diff5(i+1,1) = diff5;
end

for i=4:length(Diff5)-1
    diff6 = str2num(Diff5(i+1,1))-str2num(Diff5(i,1));
    Diff6(i,1) = diff6;
end

Diff1(length(Diff1)+1,1) = N;

Diff2(length(Diff2)+1,1) = N;

Diff3(length(Diff3)+1,1) = N;

Diff4(length(Diff4)+1,1) = N;
Diff4(length(Diff4)+1,1) = N;

Diff5(length(Diff5)+1,1) = N;
Diff5(length(Diff5)+1,1) = N;

Diff6(length(Diff6)+1,1) = N;
Diff6(length(Diff6)+1,1) = N;
Diff6(length(Diff6)+1,1) = N;

T = table(Diff1,Diff2,Diff3,Diff4,Diff5,Diff6);
disp(T)

a0 = I(1,1);
a1 = Diff1(1,1)/h;
a2 = str2num(Diff2(2,1))/(2*h^2);
a3 = str2num(Diff3(3,1))/(6*h^3);
a4 = str2num(Diff4(3,1))/(24*h^4);
a5 = str2num(Diff5(4,1))/(120*h^5);
a6 = str2num(Diff6(4,1))/(720*h^6);

disp("a0 = " + num2str(a0))
disp("a1 = " + num2str(a1))
disp("a2 = " + num2str(a2))
disp("a3 = " + num2str(a3))
disp("a4 = " + num2str(a4))
disp("a5 = " + num2str(a5))
disp("a6 = " + num2str(a6))

Eq = a0+a1*(x-t(1,1))+a2*(x-t(1,1))*(x-t(2,1))+a3*(x-t(1,1))*(x-t(2,1))*(x-t(3,1))+a4*(x-t(1,1))*(x-t(2,1))*(x-t(3,1))*(x-t(4,1))+a5*(x-t(1,1))*(x-t(2,1))*(x-t(3,1))*(x-t(4,1))*(x-t(5,1))+a6*(x-t(1,1))*(x-t(2,1))*(x-t(3,1))*(x-t(4,1))*(x-t(5,1))*(x-t(6,1));

fprintf("\nP(x) = ")
disp(simplify(Eq))

fplot(Eq,[-5,5])
grid

xpoint = input("Enter the subtitution point x = ");
fprintf("\nAt x = %g\n",xpoint)
ypoint = double(subs(Eq,x,xpoint));
fprintf("P(%g) = %g \n",xpoint,ypoint)

close all;
fplot(Eq,[-5,5])
hold;
scatter(xpoint,ypoint,'ko','markerfacecolor','black')
grid

File: Python Codes/Newton Gregory.py
#Newton Gregory
#Written by Yusuf Hani Almoadhen

import pandas as pd
from sympy import *

t = [0,1,2,3,4,5,6]
I = [176,185,194,203,212,220,229]

diff1_col=[]
diff2_col=['']
diff3_col=['','']
diff4_col=['','']
diff5_col=['','','']
diff6_col=['','','']

h = t[1]-t[0]

for i in range(0,len(I)-1):
    diff1 = I[i+1]-I[i]
    diff1_col.append(diff1)

for i in range(0,len(diff1_col)-1):
    diff2 = diff1_col[i+1]-diff1_col[i]
    diff2_col.append(diff2)

for i in range(1,len(diff2_col)-1):
    diff3 = diff2_col[i+1]-diff2_col[i]
    diff3_col.append(diff3)

for i in range(2,len(diff3_col)-1):
    diff4 = diff3_col[i+1]-diff3_col[i]
    diff4_col.append(diff4)

for i in range(2,len(diff4_col)-1):
    diff5 = diff4_col[i+1]-diff4_col[i]
    diff5_col.append(diff5)

for i in range(3,len(diff5_col)-1):
    diff6 = diff5_col[i+1]-diff5_col[i]
    diff6_col.append(diff6)

diff1_col.append('')
diff2_col.append('')
diff3_col.append('')
diff4_col.extend(['',''])
diff5_col.extend(['',''])
diff6_col.extend(['','',''])

table = pd.DataFrame({
    't': t,
    'I': I,
    '1st': diff1_col,
    '2nd': diff2_col,
    '3rd': diff3_col,
    '4th': diff4_col,
    '5th': diff5_col,
    '6th': diff6_col
})

a0 = float("{:.3g}".format(I[0]))
a1 = float("{:.3g}".format(diff1_col[0]/h))
a2 = float("{:.3g}".format(diff2_col[1]/(2*h**2)))
a3 = float("{:.3g}".format(diff3_col[2]/(6*h**3)))
a4 = float("{:.3g}".format(diff4_col[2]/(24*h**4)))
a5 = float("{:.3g}".format(diff5_col[3]/(120*h**5)))
a6 = float("{:.3g}".format(diff6_col[3]/(720*h**6)))

print("h =",h,'\n')
print(table.to_markdown(index=False),'\n')
print("a0 =",a0)
print("a1 =",a1)
print("a2 =",a2)
print("a3 =",a3)
print("a4 =",a4)
print("a5 =",a5)
print("a6 =",a6,'\n')

x = Symbol('x')
Eq = a0 + a1*(x-t[0]) + a2*(x-t[0])*(x-t[1]) + a3*(x-t[0])*(x-t[1])*(x-t[2]) + a4*(x-t[0])*(x-t[1])*(x-t[2])*(x-t[3]) + a5*(x-t[0])*(x-t[1])*(x-t[2])*(x-t[3])*(x-t[4]) + a6*(x-t[0])*(x-t[1])*(x-t[2])*(x-t[3])*(x-t[4])*(x-t[5])
print("P(x) = ",simplify(Eq))

print("At x = 0.2")
print("y =","{:.5g}".format(float(Eq.subs({x:0.2}))))
